### PR TITLE
Fix 'do_xcom_push' and 'get_logs' functionality for KubernetesJobOperator

### PR DESCRIPTION
**Pull request by MaksYermak**
In this PR I have fixed usage for `do_xcom_push` and `get_logs` functionality for `KubernetesJobOperator` in deferrable and non-deferrable modes.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by VladaZakharova** [comment]
Hi @romsharon98, @amoghrajesh @potiuk @eladkal  !
Can you please check changes here?

**Review by potiuk — APPROVED** [review]

**Comment by ccardi** [comment]
Not working on GKE Autopilot. Need to set get_logs=False.
Otherwise: pod_name is not defined.

----------
ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/python3.11/lib/python3.11/site-packages/airflow/models/taskinstance.py", line 465, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python3.11/lib/python3.11/site-packages/airflow/models/taskinstance.py", line 432, in _execute_callable
    return execute_callable(context=context, **execute_callable_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python3.11/lib/python3.11/site-packages/airflow/models/baseoperator.py", line 1724, in resume_execution
    return execute_callable(context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/gcs/dags/dag_gke_pod-simple-pi-joboperator.py", line 30, in execute_complete
    super().execute_complete(context, event)
  File "/opt/python3.11/lib/python3.11/site-packages/airflow/providers/cncf/kubernetes/operators/job.py", line 233, in execute_complete
    pod_name = event["pod_name"]
 ~~~~~^^^^^^^^^^^^
KeyError: 'pod_name'

File: airflow/providers/cncf/kubernetes/operators/job.py [hunk 1]
```diff
@@ -159,34 +166,59 @@ def execute(self, context: Context):
         ti.xcom_push(key="job_name", value=self.job.metadata.name)
         ti.xcom_push(key="job_namespace", value=self.job.metadata.namespace)
 
+        if self.pod is None:
+            self.pod = self.get_or_create_pod(  # must set `self.pod` for `on_kill`
+                pod_request_obj=self.pod_request_obj,
+                context=context,
+            )
```
Comment (NilsIrl) [inline airflow/providers/cncf/kubernetes/operators/job.py:173]: How does this work if the pod wasn't already created by the time this is reached? Won't 2 pods eventually get created (when the job eventually creates the pod)?

Comment (steinwaywhw) [reply airflow/providers/cncf/kubernetes/operators/job.py:173]: I have the same question. I thought the K8s Job will manage the pod creations, why would we need creating the pod manually again? I'm actually running into bugs where I set parallelism to more than 1 and this line triggers self.find_pods() which triggers https://github.com/apache/airflow/blob/56fbe90a8d0b56558c02d75f4ac5852e041cb058/providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/operators/pod.py#L563

Comment (amoghrajesh) [reply airflow/providers/cncf/kubernetes/operators/job.py:173]: @steinwaywhw can you explain your problem more? Would you like to take a stab at fixing it?

Comment (MaksYermak) [reply airflow/providers/cncf/kubernetes/operators/job.py:173]: Hello @steinwaywhw, 
It's a known issue https://github.com/apache/airflow/issues/44994 and the fix was already prepared by this PR https://github.com/apache/airflow/pull/49899

**Pull request merged**